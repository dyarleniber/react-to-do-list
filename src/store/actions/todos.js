import * as types from '../constants/ActionTypes';

export const addTodo = (text) => ({
  type: types.ADD_TODO,
  payload: { text },
});

export const toggleTodo = (id) => ({
  type: types.TOGGLE_TODO,
  payload: { id },
});

export const removeTodo = (id) => ({
  type: types.REMOVE_TODO,
  payload: { id },
});
