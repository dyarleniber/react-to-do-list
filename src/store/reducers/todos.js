import * as types from '../constants/ActionTypes';

const INITIAL_STATE = [];

export default function todos(state = INITIAL_STATE, action) {
  switch (action.type) {
    case types.ADD_TODO:
      return [...state, {
        id: Math.random(),
        text: action.payload.text,
        completed: false,
      }];
    case types.TOGGLE_TODO:
      return state.map((todo) => (
        todo.id === action.payload.id
          ? { ...todo, completed: !todo.completed }
          : todo
      ));
    case types.REMOVE_TODO:
      return state.filter((todo) => todo.id !== action.payload.id);
    default:
      return state;
  }
}
