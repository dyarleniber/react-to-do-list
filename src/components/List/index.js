import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Table from 'react-bootstrap/Table';
import Button from 'react-bootstrap/Button';
import Form from 'react-bootstrap/Form';

import { toggleTodo, removeTodo } from '../../store/actions/todos';

const List = () => {
  const todos = useSelector((state) => state.todos);
  const dispatch = useDispatch();

  if (!todos.length) {
    return (
      <div className="bg-light text-center">
        <small>
          No records found
        </small>
      </div>
    );
  }

  return (
    <Table hover className="bg-white">
      <tbody>
        { todos.map((todo) => {
          const { id, text, completed } = todo;
          return (
            <tr key={`rowTodo${id}`} id={`rowTodo${id}`}>
              <td className="align-middle">
                <Form.Check custom type="checkbox" id={`todo${id}`}>
                  <Form.Check.Input type="checkbox" isValid />
                  {completed
                    ? (
                      <>
                        <Form.Check.Label
                          className="labelTodo text-success"
                          onClick={() => dispatch(toggleTodo(id))}
                          tabIndex="0"
                          role="button"
                        >
                          {text}
                        </Form.Check.Label>
                        <Form.Control.Feedback className="text-success">
                          <small>
                            &#10003;
                            {' '}
                            Completed
                          </small>
                        </Form.Control.Feedback>
                      </>
                    )
                    : (
                      <>
                        <Form.Check.Label
                          className="labelTodo text-dark"
                          onClick={() => dispatch(toggleTodo(id))}
                          tabIndex="0"
                          role="button"
                        >
                          {text}
                        </Form.Check.Label>
                        <Form.Control.Feedback className="text-warning">
                          <small>
                            &#33;
                            {' '}
                            Pending
                          </small>
                        </Form.Control.Feedback>
                      </>
                    )}
                </Form.Check>
              </td>
              <td className="text-right">
                <Button
                  className="removeTodoButton"
                  variant="danger"
                  size="sm"
                  type="button"
                  onClick={() => dispatch(removeTodo(id))}
                >
                  <small>REMOVE</small>
                </Button>
              </td>
            </tr>
          );
        })}
      </tbody>
    </Table>
  );
};

export default List;
