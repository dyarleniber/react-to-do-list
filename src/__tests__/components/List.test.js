import React from 'react';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import List from '../../components/List';
import { toggleTodo, removeTodo } from '../../store/actions/todos';

const middlewares = [];
const mockStore = configureStore(middlewares);

describe('List component', () => {
  it('should render without crashing', () => {
    let wrapper;
    let initialState;
    let store;

    initialState = { todos: [] };
    store = mockStore(initialState);
    wrapper = mount(
      <Provider store={store}>
        <List />
      </Provider>,
    );
    expect(wrapper).toMatchSnapshot();

    initialState = {
      todos: [
        {
          id: 1,
          text: 'Todo 1',
          completed: true,
        },
        {
          id: 2,
          text: 'Todo 2',
          completed: false,
        },
      ],
    };
    store = mockStore(initialState);
    wrapper = mount(
      <Provider store={store}>
        <List />
      </Provider>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should toggle and remove todos', () => {
    const initialState = {
      todos: [
        {
          id: 1,
          text: 'Todo 1',
          completed: true,
        },
        {
          id: 2,
          text: 'Todo 2',
          completed: false,
        },
      ],
    };
    const store = mockStore(initialState);
    const wrapper = mount(
      <Provider store={store}>
        <List />
      </Provider>,
    );

    let actions;
    let expectedPayload;

    wrapper.find('[id="rowTodo1"] .labelTodo').first().simulate('click');
    actions = store.getActions();
    expectedPayload = toggleTodo(1);
    expect(actions).toEqual([expectedPayload]);
    store.clearActions();

    wrapper.find('[id="rowTodo2"] .labelTodo').first().simulate('click');
    actions = store.getActions();
    expectedPayload = toggleTodo(2);
    expect(actions).toEqual([expectedPayload]);
    store.clearActions();

    wrapper.find('[id="rowTodo1"] .removeTodoButton').first().simulate('click');
    actions = store.getActions();
    expectedPayload = removeTodo(1);
    expect(actions).toEqual([expectedPayload]);
    store.clearActions();

    wrapper.find('[id="rowTodo2"] .removeTodoButton').first().simulate('click');
    actions = store.getActions();
    expectedPayload = removeTodo(2);
    expect(actions).toEqual([expectedPayload]);
    store.clearActions();
  });
});
