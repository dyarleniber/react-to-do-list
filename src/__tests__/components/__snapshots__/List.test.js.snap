// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component should render without crashing 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <List>
    <div
      className="bg-light text-center"
    >
      <small>
        No records found
      </small>
    </div>
  </List>
</Provider>
`;

exports[`List component should render without crashing 2`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <List>
    <ForwardRef
      className="bg-white"
      hover={true}
    >
      <table
        className="bg-white table table-hover"
      >
        <tbody>
          <tr
            id="rowTodo1"
            key="rowTodo1"
          >
            <td
              className="align-middle"
            >
              <FormCheck
                custom={true}
                disabled={false}
                id="todo1"
                inline={false}
                isInvalid={false}
                isValid={false}
                title=""
                type="checkbox"
              >
                <div
                  className="custom-control custom-checkbox"
                >
                  <FormCheckInput
                    isValid={true}
                    type="checkbox"
                  >
                    <input
                      className="custom-control-input is-valid"
                      id="todo1"
                      type="checkbox"
                    />
                  </FormCheckInput>
                  <FormCheckLabel
                    className="labelTodo text-success"
                    onClick={[Function]}
                    role="button"
                    tabIndex="0"
                  >
                    <label
                      className="labelTodo text-success custom-control-label"
                      htmlFor="todo1"
                      onClick={[Function]}
                      role="button"
                      tabIndex="0"
                    >
                      Todo 1
                    </label>
                  </FormCheckLabel>
                  <Feedback
                    className="text-success"
                    type="valid"
                  >
                    <div
                      className="text-success valid-feedback"
                    >
                      <small>
                        âœ“
                         
                        Completed
                      </small>
                    </div>
                  </Feedback>
                </div>
              </FormCheck>
            </td>
            <td
              className="text-right"
            >
              <Button
                active={false}
                className="removeTodoButton"
                disabled={false}
                onClick={[Function]}
                size="sm"
                type="button"
                variant="danger"
              >
                <button
                  className="removeTodoButton btn btn-danger btn-sm"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  <small>
                    REMOVE
                  </small>
                </button>
              </Button>
            </td>
          </tr>
          <tr
            id="rowTodo2"
            key="rowTodo2"
          >
            <td
              className="align-middle"
            >
              <FormCheck
                custom={true}
                disabled={false}
                id="todo2"
                inline={false}
                isInvalid={false}
                isValid={false}
                title=""
                type="checkbox"
              >
                <div
                  className="custom-control custom-checkbox"
                >
                  <FormCheckInput
                    isValid={true}
                    type="checkbox"
                  >
                    <input
                      className="custom-control-input is-valid"
                      id="todo2"
                      type="checkbox"
                    />
                  </FormCheckInput>
                  <FormCheckLabel
                    className="labelTodo text-dark"
                    onClick={[Function]}
                    role="button"
                    tabIndex="0"
                  >
                    <label
                      className="labelTodo text-dark custom-control-label"
                      htmlFor="todo2"
                      onClick={[Function]}
                      role="button"
                      tabIndex="0"
                    >
                      Todo 2
                    </label>
                  </FormCheckLabel>
                  <Feedback
                    className="text-warning"
                    type="valid"
                  >
                    <div
                      className="text-warning valid-feedback"
                    >
                      <small>
                        !
                         
                        Pending
                      </small>
                    </div>
                  </Feedback>
                </div>
              </FormCheck>
            </td>
            <td
              className="text-right"
            >
              <Button
                active={false}
                className="removeTodoButton"
                disabled={false}
                onClick={[Function]}
                size="sm"
                type="button"
                variant="danger"
              >
                <button
                  className="removeTodoButton btn btn-danger btn-sm"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  <small>
                    REMOVE
                  </small>
                </button>
              </Button>
            </td>
          </tr>
        </tbody>
      </table>
    </ForwardRef>
  </List>
</Provider>
`;
